src/ocf_composite_volume.c

```diff
@@ -0,0 +1,357 @@
+/*
+ * Copyright(c) 2022 Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "ocf/ocf.h"
+#include "ocf_priv.h"
+#include "ocf_core_priv.h"
+#include "ocf_io_priv.h"
+#include "metadata/metadata.h"
+#include "engine/cache_engine.h"
+#include "utils/utils_user_part.h"
+#include "ocf_request.h"
+#include "ocf_composite_volume_priv.h"
+
+#define OCF_COMPOSITE_VOLUME_MEMBERS_MAX 16
+
+struct ocf_composite_volume {
+	uint8_t members_cnt;
+	struct {
+		struct ocf_volume volume;
+		void *volume_params;
+	} member[OCF_COMPOSITE_VOLUME_MEMBERS_MAX];
+	uint64_t lookup[OCF_COMPOSITE_VOLUME_MEMBERS_MAX];
+	uint64_t length;
+	unsigned max_io_size;
+};
+
+struct ocf_composite_volume_io {
+	struct ocf_io *io[OCF_COMPOSITE_VOLUME_MEMBERS_MAX];
+	ctx_data_t *data;
+	uint8_t begin_member;
+	uint8_t end_member;
```

Comment:
Don't you think `end_member` is a little confusing? It stores an information about the number of volumes which are affected by an IO, not the id of the last volume. How about `member_count`?

Reply:
Not really. It actually is an index. See how `begin_member` is used. We always iterate **between** `begin_member` **and** `end_member`, not **starting from** `begin_member` **by** `end_member` **count**.